### PR TITLE
[j2k] Modify the Java Search for Final Variables for Kotlin Conversion

### PR DESCRIPTION
**Pull request by jocelynluizzi13**
As discussed in #2743 , the java plugin to search for "final" variables doesn't quite work for J2K conversions since Kotlin mutability is slightly less strict. As such, making some changes and adding new functions in the Java plugin to leverage the existing infrastructure in the Kotlin J2K plugin.

Here are the separated out Java changes alongside new test cases.

@abelkov @darthorimar @ermattt @amaembo

### CODE REVIEW

File: java/java-analysis-impl/src/com/intellij/codeInspection/localCanBeFinal/LocalCanBeFinal.java [hunk 1]
```diff
@@ -351,6 +317,197 @@ public void visitForStatement(@NotNull PsiForStatement statement) {
...
 
+  /**
+  * This method is to be used in the Java 2 Kotlin conversion to determine
+  * local variables that should be defined with "val" (implying immutability)
+  *<p>
+  * @param body the block of code to search
+  * @return a set of kotlin immutable PsiVariables
+   */
+  @Nullable
+  public Set<PsiVariable> getKotlinImmutableLocalVariablesInBlock(final PsiCodeBlock body) {
```
Comment (amaembo) [inline java/java-analysis-impl/src/com/intellij/codeInspection/localCanBeFinal/LocalCanBeFinal.java:328]: Is it possible to keep Kotlin-specific behavior in the Kotlin plugin? It looks out of scope of Java plugin, and an additional burden for Java team to maintain this code.

Comment (jocelynluizzi13) [reply java/java-analysis-impl/src/com/intellij/codeInspection/localCanBeFinal/LocalCanBeFinal.java]: works for me, I moved it back but left the refactor of Java so that I could reuse some code @amaembo

File: java/java-analysis-impl/src/com/intellij/codeInspection/localCanBeFinal/LocalCanBeFinal.java [hunk 2]
```diff
@@ -351,6 +317,45 @@ public void visitForStatement(@NotNull PsiForStatement statement) {
     return problems;
   }
 
+  @Nullable
+  public static ControlFlow getControlFlow(final PsiCodeBlock body) {
```
Comment (amaembo) [inline java/java-analysis-impl/src/com/intellij/codeInspection/localCanBeFinal/LocalCanBeFinal.java:321]: Please document the public method. What kind of control flow is returned? How it differs from existing methods in `ControlFlowFactory`?

File: java/java-analysis-impl/src/com/siyeh/ig/psiutils/FinalUtils.java [hunk 3]
```diff
@@ -26,28 +26,44 @@ public static boolean canBeFinal(@NotNull PsiVariable variable) {
...
-        if (innerScope != null && innerScope != ((PsiField)variable).getContainingClass()) return false;
-      }
-      return HighlightControlFlowUtil.checkFinalVariableMightAlreadyHaveBeenAssignedTo(variable, ref, finalVarProblems) == null;
+      return checkElementDoesNotViolateFinality(e, variable, uninitializedVarProblems, finalVarProblems);
     };
     return PsiTreeUtil.processElements(scope, elementDoesNotViolateFinality);
   }
+
+  /*
+  * this method is used in the Kotlin plugin J2K process
```
Comment (amaembo) [inline java/java-analysis-impl/src/com/siyeh/ig/psiutils/FinalUtils.java:46]: Please write a javadoc concentrating on what this new public method does and what its parameters mean instead of where it's used. It's not so necessary to say where it's used. Thank you.

**Review by amaembo — CHANGES_REQUESTED** [review]
The code looks good. Please update the documentation, and I'll merge. Thank you.

File: java/java-analysis-impl/src/com/siyeh/ig/psiutils/FinalUtils.java (hunk 3 above)
Comment (jocelynluizzi13) [reply java/java-analysis-impl/src/com/siyeh/ig/psiutils/FinalUtils.java]: updated! thank you!

**Comment by amaembo** [comment]
Pushed, thanks. 0e11a2c3d7ecfff4ea84829d5c2ccd7fcca0e4ff
